storage/storage-interface/src/lib.rs

```diff
@@ -534,6 +530,67 @@ pub trait DbWriter: Send + Sync {
         latest_in_memory_state: StateDelta,
         state_updates_until_last_checkpoint: Option<ShardedStateUpdates>,
         sharded_state_cache: Option<&ShardedStateCache>,
+    ) -> Result<()> {
+        // For reconfig suffix.
+        if ledger_info_with_sigs.is_none() && txns_to_commit.is_empty() {
+            return Ok(());
+        }
+
+        if !txns_to_commit.is_empty() {
+            self.pre_commit_ledger(
+                txns_to_commit,
+                first_version,
+                base_state_version,
+                sync_commit,
+                latest_in_memory_state,
+                state_updates_until_last_checkpoint,
+                sharded_state_cache,
+            )?;
+        }
+        let version_to_commit = if let Some(ledger_info_with_sigs) = ledger_info_with_sigs {
+            ledger_info_with_sigs.ledger_info().version()
+        } else {
+            // here txns_to_commit is known to be non-empty
+            first_version + txns_to_commit.len() as u64 - 1
+        };
+        self.commit_ledger(
+            version_to_commit,
+            ledger_info_with_sigs,
+            Some(txns_to_commit),
+        )
+    }
+
+    /// Optimistically persist transactions to the ledger.
+    ///
+    /// Called by consensus to pre-commit blocks before execution result is agreed on by the
+    /// validators.
+    ///
+    ///   If these blocks are later confirmed to be included in the ledger, commit_ledger should be
+    ///       called with a `LedgerInfoWithSignatures`.
+    ///   If not, the consensus needs to panic, resulting in a reboot of the node where the DB will
+    ///       truncate the unconfirmed data.
+    fn pre_commit_ledger(
+        &self,
+        txns_to_commit: &[TransactionToCommit],
+        first_version: Version,
+        base_state_version: Option<Version>,
+        sync_commit: bool,
+        latest_in_memory_state: StateDelta,
+        state_updates_until_last_checkpoint: Option<ShardedStateUpdates>,
+        sharded_state_cache: Option<&ShardedStateCache>,
+    ) -> Result<()> {
+        unimplemented!()
+    }
+
+    /// Commit pre-committed transactions to the ledger.
+    ///
+    /// If a LedgerInfoWithSigs is provided, both the "synced version" and "committed version" will
+    /// advance, otherwise only the synced version will advance.
+    fn commit_ledger(
+        &self,
+        version: Version,
+        ledger_info_with_sigs: Option<&LedgerInfoWithSignatures>,
+        txns_to_commit: Option<&[TransactionToCommit]>,
```

Comment:
is this None for pre-commit case?

Reply:
Yes. For state sync this can be passed in so we can save scanning the write set CF for the table info indexer (gonna be removed anyway)

Reply:
it won't work when CO enabled?

Reply:
why? you can always pass in None. It'll scan the for the write sets.

Reply:
how slow is that?

Reply:
1. Only the indexer fullnodes have that enabled, I don't think committing is the bottleneck there.
1. We will migrate away anyways. In fact I suspect we already did. Was confirming with the indexer team on slack today.